Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs

```diff
@@ -5,229 +5,267 @@
+            };
 
-		[Fact]
-		public async void FailOnStateMismatchInInitialRequest()
-		{
-			OidcClient client = new OidcClient(new OktaConfig("testoktaid", "https://dev-00000.oktapreview.com", "com.test:/redirect", "com.test:/logout"));
+            client.SetHttpMock(mockHttpClient);
 
-			client.OnLaunchBrowser = new Action<string>(url =>
-			{
-				Assert.False(
-					OidcClient.CaptureRedirectUrl(new Uri(client.Config.RedirectUri + "?code=12345&state=fake_state")));
-			});
+            client.OnLaunchBrowser = new Action<string>(url =>
+            {
+                OidcClient.CaptureRedirectUrl(new Uri(client.Config.RedirectUri + "?code=12345&state=" + client.State_Internal));
+            });
 
-			await Task.WhenAny(client.SignInWithBrowserAsync(), Task.Delay(1000));
-		}
+            await Assert.ThrowsAsync<OAuthException>(() => client.SignInWithBrowserAsync());
+        }
+
+        [Fact]
+        public async void FailGracefullyOnHttpErrorInAccessTokenRequest()
+        {
+            OidcClient client = new OidcClient(new OktaConfig("testoktaid", "https://dev-00000.oktapreview.com", "com.test:/redirect", "com.test:/logout"));
+
+            HttpMessageHandlerMock mockHttpClient = new HttpMessageHandlerMock();
+            mockHttpClient.Responder = (request) =>
+            {
+                string url = request.Item1;
+                Dictionary<string, string> data = request.Item2;
+
+                return new Tuple<System.Net.HttpStatusCode, string>(
+                    System.Net.HttpStatusCode.Forbidden,
+                    @"{ ""error"": ""not_authorized"", ""token_type"": ""testing""}");
+            };
+
+            client.SetHttpMock(mockHttpClient);
+
+            client.OnLaunchBrowser = new Action<string>(url =>
+            {
+                OidcClient.CaptureRedirectUrl(new Uri(client.Config.RedirectUri + "?code=12345&state=" + client.State_Internal));
+            });
+
+            await Assert.ThrowsAsync<OAuthException>(() => client.SignInWithBrowserAsync());
+        }
 
 		[Fact]
-		public async void FailOnErrorInInitialRequest()
+		public void SetStateCodeVerifierAndChallengeOnSignOut()
 		{
 			OidcClient client = new OidcClient(new OktaConfig("testoktaid", "https://dev-00000.oktapreview.com", "com.test:/redirect", "com.test:/logout"));
-
-			client.OnLaunchBrowser = new Action<string>(url =>
-			{
-				Assert.True(
-					OidcClient.CaptureRedirectUrl(new Uri(client.Config.RedirectUri + "?error=test_error&state=" + client.State_Internal)));
-			});
-
-			await Assert.ThrowsAsync<OAuthException>(() => client.SignInWithBrowserAsync());
+			Assert.True(string.IsNullOrEmpty(client.State_Internal));
+			Assert.True(string.IsNullOrEmpty(client.CodeChallenge_Internal));
+			Assert.True(string.IsNullOrEmpty(client.CodeVerifier_Internal));
+			
+			client.SignOutOfOktaAsync(new OktaState("testAccessToken", "testTokenType"));
+			
+			Assert.False(string.IsNullOrEmpty(client.State_Internal));
```

Comment:
I'm not quite following this test. What exactly are we testing here?

Reply:
This is making sure the internal state of the client gets set by the call to signout.  The State, CodeChallenge and CodeVerifier are private to the client but exposed as "_Internal" for testing.  Those values are needed for the background redirect flows.